### DISCUSSION TITLE
How to configure an asset profile with a scraper configuration?

### DISCUSSION THREAD

**Discussion by antanisblinda** [ghostfolio/ghostfolio#2819] [Q&A]
Hey there!
following a discussion on twitter, opening this thread.
I was looking at the premium option for ghostfolio and started with the demo to see how it would work for me.
I noticed that certain (italian) govt bonds cannot be added (example, ISIN IT0001174611, available [here](https://www.borsaitaliana.it/borsa/obbligazioni/mot/btp/lista.html?lang=en)).
I understand that i could track it as "valuable" but of course it would mean no automated updates of market value/expiration/etc. 
Another option would be a scraper, for example from the URL noted above.
- any hints or documentation reference on how this can be done?
- is this only an option for the self-hosted version or also for the premium tier? I was really looking for a smooth transition without hacking my way into it.

thank you!

**Comment by dtslvr** [comment]
Hi @antanisblinda

In the self-hosted version of Ghostfolio, you can create custom asset profiles using a scraper configuration. The market data is then crawled periodically. To achieve that, follow these steps:

1. Open the _Admin Control_ panel
2. Go to _Market Data_
3. Create an asset profile (_Add manually..._)
4. _Edit_ your asset profile
5. Add the following snippet to _Scraper Configuration_:
```
{"selector":".summary-value strong","url":"https://www.borsaitaliana.it/borsa/obbligazioni/mot/btp/scheda/IT0001174611.html?lang=en"}
```
6. _Save_ the asset profile

**Reply by davidecorradini** [reply]
How can I add the country(ies) to a MANUAL asset? For example, in this case, it would be 100% Italy. I tried several formats for the "countries" field in the asset, such as `["IT"]`, `["ITALY"]`, `[{"IT": 1}]`, but the input is not accepted.

**Reply by dtslvr** [reply]
This is the format: `[{"code": "IT", "weight": 1}]`

**Reply by gstr3745** [reply]
Thank you @dtslvr ! I've been searching for this since a while!
Via trials and errors I found out that syntax for sectors is quite similar `[{"name":"Treasury","weight":"1"}]`

Are there any lists of accepted values?

**Reply by dtslvr** [reply]
There is currently no full list of accepted values. However, here are the values extracted from an _FTSE All-World_ ETF that might be helpful (not claiming to be complete).

- Chemicals
- Communication
- Communication Services
- Consumer Cyclical
- Consumer Staples
- Electric Appliances
- Energy
- Financials
- Healthcare
- Industrials
- Information & Communication
- Materials
- Real Estate
- Services
- Technology
- Telecommunications
- Transportation Equipment
- Utilities

**Reply by gstr3745** [reply]
Thank you very much @dtslvr !

**Reply by diamant-x** [reply]
Where is the 'scrapper configuration' documentation? To understand what parammeters are available to be used, and how do they impact on the end result. Thanks.

**Reply by ziritione85** [reply]
> Thank you @dtslvr ! I've been searching for this since a while! Via trials and errors I found out that syntax for sectors is quite similar `[{"name":"Treasury","weight":"1"}]`
> 
> Are there any lists of accepted values?

How do you add more than one line, I mean more countries and sectors?

**Reply by gstr3745** [reply]
For countries: `[{"code":"FR","weight":0.25},{"code":"DE","weight":0.25},{"code":"NL","weight":0.25},{"code":"ES","weight":0.25}]`
For sectors: `[{"name":"Technology","weight":0.25},{"name":"Healthcare","weight":0.25},{"name":"Financial Services","weight":0.25},{"name":"Industrials","weight":0.25}]`

**Comment by zaromarco** [comment]
Hi , i'm trying to get data with scraper from https://live.euronext.com/nb/product/structured-products/IT0005333742-ETLX/market-information
but with the example provided everytime i get parse error , also on other site
is there a guide to implement scraper - sector - symbol mapping and a guide to fully use the potential of market data?

**Reply by gstr3745** [reply]
Hi zaromarco, I could get it working with this scraper configuration:
`{"selector":"div.-pb:nth-child(3) > div:nth-child(1) > table:nth-child(1) > tbody:nth-child(1) > tr:nth-child(5) > td:nth-child(2) > span:nth-child(1)","url":"https://www.borsaitaliana.it/borsa/cw-e-certificates/eurotlx/scheda/IT0005333742.html?lang=en"}`

Euronext.com seems to throw this error:
`XHRPOST
Status: 400 Bad Request
Referrer Policy: strict-origin-when-cross-origin`

**Reply by zaromarco** [reply]
hi is there a guide to understand what you have done? because i have tested on borsaitaliana other italian isin code , not found in yahoo finance , and with your code no ones works, here is the list of isin:
IT0005532723
IT0004513641
IT0005547408
IT0005557084
IT0005584302
IT0005555963
IT0005086878

this is: LU0290358497 is an xtracker etf and i don't know why yahoo finance is not able to find it...on borsaitalia is there...

many thank's

**Reply by zaromarco** [reply]
have you used some kind of software to analyze webpage and build the scraper code?

**Reply by gstr3745** [reply]
The scraper is quite simple:
`{"selector":"SEL_VALUE","url":"ADDRESS"}`

To get the right data, first of all you've got to get the page you could find the price on: i.e. for IT0005086878 you've got to change `ADDRESS` with `https://www.borsaitaliana.it/borsa/obbligazioni/eurotlx/scheda/IT0005086878.html?lang=en`

To set the `SEL_VALUE`, you've got to inspect the page and copy the CSS Selector: right click on the data field and then click on "inspect"/"analyze" or something like that, depending on your browser.
You should get to the page html source, right click on the proper line, then "Copy css selector"
For the example above, it would get the value `div.-prl:nth-child(1) > table:nth-child(1) > tbody:nth-child(1) > tr:nth-child(5) > td:nth-child(2) > span:nth-child(1)` assuming you're analyzing the close price value.

This way, the scraper configuration should be:
`{"selector":"div.-prl:nth-child(1) > table:nth-child(1) > tbody:nth-child(1) > tr:nth-child(5) > td:nth-child(2) > span:nth-child(1)","url":"https://www.borsaitaliana.it/borsa/obbligazioni/eurotlx/scheda/IT0005086878.html?lang=en"}`

~~I would assume that, within the borsaitaliana.it domain, all you must change is the url value.~~
You should check on every stock, since the fields and their positions are not constant between different products.

**Reply by zaromarco** [reply]
Many many thank's for the explanation , it works like a charm in firefox!

**Reply by zaromarco** [reply]
What kind of software use ghostfolio to use the scraper script?
I'm thinking of using the same for a standalone project

**Reply by dtslvr** [reply]
> What kind of software use ghostfolio to use the scraper script?

Check `scrape()` in [ManualService](https://github.com/ghostfolio/ghostfolio/blob/main/apps/api/src/services/data-provider/manual/manual.service.ts) (https://github.com/cheeriojs/cheerio).

**Reply by SpencerSmithSite** [reply]
Can't seem to get this to work with any other data sources. as an example, when trying to get the market price for https://finance.yahoo.com/quote/VOO
I tried the following in Admin Control | Market Data | Manual | scraper configuration:
```
{"selector":"li.svelte-tx3nkj:nth-child(1) > span:nth-child(2) > fin-streamer:nth-child(1)","url":"https://finance.yahoo.com/quote/VOO"}
```
It returns the following error:
"Error: Could not parse the current market price"
The same is the case with every data source I've tried. Am I missing something?

**Reply by zaromarco** [reply]
why you need to use the manual market data?
this thicker is correctly found when you add a new activity
i'm asking this also because in ghostfolio the default market data is yahoo finance

**Reply by SpencerSmithSite** [reply]
The only ticker I'm getting returned is "vanguard-sp-500-etf-tokenized-stock-defichain", which is not the true etf that I'm looking for. I can't seem to pull other financial tickers either, only cryptos

**Reply by zaromarco** [reply]
very strange...i'm using the latest version self hosted and when i create a new activity and in the name i type VOO i get 6 different etf which one is the same on yahoo finance page

**Reply by SpencerSmithSite** [reply]
I am also self-hosted, version 2.61.0

**Reply by zaromarco** [reply]
that version is from march '24 , i'm using 2.77.1
maybe is worth try to upgrade and see what will show

**Reply by gstr3745** [reply]
Market data fetch is broken for previous versions (#3276 , corrected with #3286 ): you should upgrade at least to 2.69.0 (if I recall it right).
Anyway, latest is working smoothely

**Comment by colorcube** [comment]
Because related:

- what data should be put into the field "Symbol Mapping"
- what does Benchmark do?
- can the imported "Historical Data" be deleted (in case of wrong data for example)?

**Reply by gstr3745** [reply]
you shuld populate "symbol mapping" to get the sector and countries data (e.g. from trackinsight)

the check on "benchmark" lets you graph the asset on your portfolio overview, as far as I understood: so you could compare easily your portfolio with the single asset or index

as far as I know, you could overwrite historical data with manual csv import, but I'm not aware if it's possible to delete previous scraped or gathered data, I'm sorry

This is what I assumed, using ghostfolio: I hope I'm not misunderstanding anything.

**Reply by colorcube** [reply]
> you shuld populate "symbol mapping" to get the sector and countries data (e.g. from trackinsight)

How? what's the format?

**Reply by colorcube** [reply]
For ISIN IE00BJ0KDQ92 the trackinsight page is
https://www.trackinsight.com/de/fund/24869934

In the source code you find the code QQQ for TrackinsightDataEnhancerService.
Which means the field "symbol mapping" should be:

`{"QQQ":"24869934"}`

But I have no idea how to trigger the service. "Gather * Data" menu options has no effect.

**Reply by gstr3745** [reply]
If I recall right, it should be:
`{"TRACKINSIGHT":"24869934"}`

I cannot try this right now on my instance, I hope this will work

**Reply by colorcube** [reply]
Right, still no luck. I'll look into it. 

Thanks for your help!

**Reply by swissbuechi** [reply]
`{"TRACKINSIGHT":"VT"}` works for me.
Clicked on the refresh profile data button and waited for about 5 min until I checked the data in the allocations screen. 👍🏼

**Reply by M4tT3d** [reply]
if I set a symbol mapping like this `{"TRACKINGSIGHT":"IWDA"}` into the configuration of market data item, I cannot see any change. From nest logs:

```
LOG [DataGatheringProcessor (GATHER_ASSET_PROFILE)] Asset profile data gathering has been started for SWDA.MI (YAHOO)
LOG [DataGatheringProcessor (GATHER_ASSET_PROFILE)] Asset profile data gathering has been completed for SWDA.MI (YAHOO)
```

Am I doing it wrong or could be a bug?

**Reply by dtslvr** [reply]
It is called `TRACKINSIGHT` , not `TRACKINGSIGHT`.

**Reply by M4tT3d** [reply]
> It is called `TRACKINSIGHT` , not `TRACKINGSIGHT`.

I made a mistake while writing the previous message, but in symbol mapping it is as you wrote it. I am trying to map a non-manual type item

**Comment by fib1985** [comment]
Hi all! How I can customize the "ETF provider" field?

**Reply by fib1985** [reply]
Nobody? I want to customize this field for the manual added Items in marker data section.

**Reply by swissbuechi** [reply]
@dtslvr 
Are the ETF providers hard-coded or is there a dynamic way to configure them?

**Comment by hvddrift** [comment]
Hi

I am trying to scrape (first attempt) from https://www.vanguard.com.au/personal/invest-with-us/fund?portId=8100&tab=prices-and-distributions 

```
{
  "url": "https://www.vanguard.com.au/personal/invest-with-us/fund?portId=8100&tab=prices-and-distributions",
  "selector": "span.text--like-h6",
  "mode": "lazy"
}
```

response: Error
Could not parse the current market price for VAN0002AU (MANUAL)

or JSON

```
{
  "url": "https://www.vanguard.com.au/personal/api/products/personal/fund/8100/prices?limit=-1",
  "selector": "$.data[0].navPrices[0].price",
  "mode": "lazy"
}
```

Error
Cannot read properties of undefined (reading 'query')

Any syntax or code suggestion would be appreciated. 

Hans

**Reply by gstr3745** [reply]
I suppose it's because of the $ char in the value field: it should be a valid number formatted value.
I'd suggest to refer to another provider (e.g. investing.com) which provide prices in plain values.
This should work:
`{"url":"https://www.investing.com/funds/au60van00022","selector":"#last_last"}`

Perhaps a more skilled user could help you further with a valid scraper selector or configuration with the url you are using 🤞

**Reply by gstr3745** [reply]
I'm sorry, I totally ignored the json scraper error you posted... I found this open issue (already solved, but waiting to be implemented): #3997 
It looks like the same error

**Reply by hvddrift** [reply]
Thank you, I spent way too long trying to get rid of the $, and even longer looking for another pricing site. This worked well. Thank you.

**Comment by cardom** [comment]
Hi @dtslvr, what the "selector" specifically is?

I'm trying to scraper the closing price in a table format published here: https://www.boerse-frankfurt.de/en/etf/wisdomtree-physical-bitcoin/price-history/historical-prices-and-volumes?currency=EUR
The data I need is provided from an application in the webpage under a table format (td, tr), and the selector I can get from the Inspection of the web page is, just for the first row (latest traded day) and its close price is:

"body > app-root > app-wrapper > div > div > div.content-wrapper > app-etp > div.row.ng-star-inserted > div > app-price-history > div.widget.app-loading-spinner-parent.d-block.ar-mt.ar-p > div > div > div.table-responsive > table > tbody > tr:nth-child(1) > td:nth-child(3)"

Trying to add this selector it does not works (and the format is clearly wrong if compared with your example and the selector I can find with the same method in the page of your example).

Could you please give us some pointer (what is the scraper used, a roughly process on how to find the selector, and so on...) which can help anyone of us to find its own scraper settings?

Hope you can find the time to help in any way you may want!

Thanks in advance

**Reply by dtslvr** [reply]
This configuration is correct:
```
{"selector":"body > app-root > app-wrapper > div > div > div.content-wrapper > app-etp > div.row.ng-star-inserted > div > app-price-history > div.widget.app-loading-spinner-parent.d-block.ar-mt.ar-p > div > div > div.table-responsive > table > tbody > tr:nth-child(1) > td:nth-child(3)","url":"https://www.boerse-frankfurt.de/en/etf/wisdomtree-physical-bitcoin/price-history/historical-prices-and-volumes?currency=EUR"}
```
It seems the data table is loaded asynchronously and the scraper isn't able to access the content. Could you try using another url?

**Reply by cardom** [reply]
Yes, should be asynchronous, since the table is offered by a backend app.
Unfortunately I have just two ETFs for which I need to import manually the data, and are both exchanged on XETRA (same URL, same backend app, same kind of table), so I don't have nothing else to try with.

I think then, for now, is not possible to use the built in scraper?

Do you think that is possible to integrate a way to manage in asynchronous the data to scrape? If it is possible, and can be done without too much effort, I would like to open a PR for it, if it's needed

**Reply by dtslvr** [reply]
A headless browser like Puppeteer might be needed for this approach, but in my opinion, it's overkill.

Maybe you could try something like this: `{"selector":".mod-ui-table td:nth-last-child(2)","url":"https://markets.ft.com/data/etfs/tearsheet/historical?s=WBIT:GER:EUR"}`

**Comment by Section82** [comment]
I'm trying to scrape the price from https://www.sgx.com/fixed-income/products/EHZB and have no clue where to start or what to enter in any of the fields. On firefox if I right click and view source and then copy CSS selector for the price I get ".sgx-price-value" so I put this in the "selector" field and https://www.sgx.com/fixed-income/products/EHZB in the URL field but I just get a "Could not parse the current market price for SGXF29028044 (MANUAL)"

What the hell am I doing wrong and how the hell do I use this feature? Does it handle logins as well? i.e. if I wanted to get a price from the FT market data using my subscription, would it work?

**Comment by solrai38** [comment]
In a similar vein, I'm struggling to figure out how the scraper configuration works, in my case I'm trying to pull a couple of crypto-asset prices from an Aussie exchange that publishes prices on https://ainsliecrypto.com.au/Dashboard as both part of a live ticker, but more sensibly a table.

Sadly, the two in question AUS and AGS) aren't listed on the more common sites such as Coingecko so I'm limited for choices of how to get updated pricing.  Any assistance in getting to the right scraper configuration would be much appreciated!

**Comment by Jalvcl** [comment]
If someone helpme (us) scrape msn finnace helps a lot because is a free mutual fund prices... i cant find another one 


https://www.msn.com/es-cl/dinero/funddetails/fi-bwhorw?id=bwhorw
